File: openclaw_safe/kernel/capability_manifest.py
"""CapabilityManifest – declares what capabilities a plugin requires."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


@dataclass
class Capability:
    """A single declared capability with its associated risk level."""

    name: str
    description: str
    risk_level: RiskLevel = RiskLevel.LOW

    def __str__(self) -> str:
        return f"Capability({self.name!r}, risk={self.risk_level.value})"


@dataclass
class CapabilityManifest:
    """The full set of capabilities declared by a plugin."""

    plugin_name: str
    version: str
    capabilities: list[Capability] = field(default_factory=list)

    def add_capability(self, capability: Capability) -> None:
        self.capabilities.append(capability)

    def get_capability(self, name: str) -> Capability | None:
        for cap in self.capabilities:
            if cap.name == name:
                return cap
        return None

    def max_risk_level(self) -> RiskLevel:
        if not self.capabilities:
            return RiskLevel.LOW
        order = [RiskLevel.LOW, RiskLevel.MEDIUM, RiskLevel.HIGH, RiskLevel.CRITICAL]
        max_idx = max(order.index(c.risk_level) for c in self.capabilities)
        return order[max_idx]

    def __str__(self) -> str:
        return (
            f"CapabilityManifest(plugin={self.plugin_name!r}, "
            f"version={self.version!r}, capabilities={len(self.capabilities)})"
        )

File: src/core/capability_manifest.ts
export enum RiskLevel {
  LOW = "low",
  MEDIUM = "medium",
  HIGH = "high",
  CRITICAL = "critical",
}

const RISK_ORDER: RiskLevel[] = [RiskLevel.LOW, RiskLevel.MEDIUM, RiskLevel.HIGH, RiskLevel.CRITICAL];

export interface Capability {
  name: string;
  description: string;
  riskLevel: RiskLevel;
}

export interface CapabilityManifest {
  pluginName: string;
  version: string;
  capabilities: Capability[];
}

export function createCapabilityManifest(pluginName: string, version: string): CapabilityManifest {
  return { pluginName, version, capabilities: [] };
}

export function addCapability(manifest: CapabilityManifest, cap: Capability): void {
  manifest.capabilities.push(cap);
}

export function getCapability(manifest: CapabilityManifest, name: string): Capability | undefined {
  return manifest.capabilities.find((c) => c.name === name);
}

export function maxRiskLevel(manifest: CapabilityManifest): RiskLevel {
  if (manifest.capabilities.length === 0) return RiskLevel.LOW;
  const maxIdx = Math.max(...manifest.capabilities.map((c) => RISK_ORDER.indexOf(c.riskLevel)));
  return RISK_ORDER[maxIdx];
}
